File: backend/app/api/routes/product.py
from fastapi import APIRouter, Depends, HTTPException

from app.database import get_db
from app.models import ProductCreate, ProductUpdate

router = APIRouter(prefix="/products")


@router.get("/all")
def get_all_products(db=Depends(get_db)):
    cursor = db.cursor(dictionary=True)

    query = "SELECT * FROM PRODUCT"
    cursor.execute(query)
    return cursor.fetchall()


@router.post("/save")
def save_product(product: ProductCreate, db=Depends(get_db)):
    product_data = (
        product.PROD_CODE,
        product.PROD_DESCRIPT,
        product.PROD_INDATE,
        product.PROD_QOH,
        product.PROD_MIN,
        product.PROD_PRICE,
        product.PROD_DISCOUNT,
        product.VEND_NUMBER,
    )

    try:
        cursor = db.cursor()
        cursor.execute(
            """
            INSERT INTO PRODUCT
            (PROD_CODE, PROD_DESCRIPT, PROD_INDATE, PROD_QOH, PROD_MIN, PROD_PRICE, PROD_DISCOUNT, VEND_NUMBER)
            VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
            """,
            product_data,
        )

        db.commit()
    except:
        raise HTTPException(status_code=500, detail="Failed to save product")


@router.put("/update/{id:path}")
def update_product(id: str, product: ProductUpdate, db=Depends(get_db)):
    product_data = (
        product.PROD_DESCRIPT,
        product.PROD_INDATE,
        product.PROD_QOH,
        product.PROD_MIN,
        product.PROD_PRICE,
        product.PROD_DISCOUNT,
        product.VEND_NUMBER,
    )
    try:
        cursor = db.cursor(dictionary=True)
        cursor.execute(
            """
            UPDATE PRODUCT
            SET PROD_DESCRIPT = %s, PROD_INDATE = %s, PROD_QOH = %s, PROD_MIN = %s, PROD_PRICE = %s, PROD_DISCOUNT = %s, VEND_NUMBER = %s
            WHERE PROD_CODE = %s
            """,
            (*product_data, id),
        )
        db.commit()
    except:
        raise HTTPException(status_code=500, detail="Failed to update product")


@router.delete("/delete/{product_id:path}", status_code=204)
def delete_product_by_id(product_id: str, db=Depends(get_db)):
    cursor = db.cursor()

    cursor.execute("""DELETE FROM LINE WHERE PROD_CODE = %s""", [product_id])

    cursor.execute(
        """
                   DELETE FROM PRODUCT WHERE PROD_CODE = %s""",
        [product_id],
    )

    db.commit()

File: frontend/app/api/product.ts
import apiFetch from "./api";

export type Product = {
  PROD_CODE: string;
  PROD_DESCRIPT: string | null;
  PROD_INDATE: Date | null;
  PROD_QOH: number | null;
  PROD_MIN: number | null;
  PROD_PRICE: number | null;
  PROD_DISCOUNT: number | null;
  VEND_NUMBER: number | null;
};

export type ProductUpdate = {
  PROD_DESCRIPT: string;
  PROD_INDATE: Date;
  PROD_QOH: number;
  PROD_MIN: number;
  PROD_PRICE: number;
  PROD_DISCOUNT: number;
  VEND_NUMBER: number;
};

export async function findAllProducts() {
  const response = await apiFetch("/products/all");

  if (!response) {
    throw new Error("Failed to retrieve products.");
  }

  const data = (await response.json()) as Product[];
  return data;
}

export async function saveProduct(newProduct: Product) {
  const response = await apiFetch("/products/save", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(newProduct),
  });

  if (!response) {
    throw new Error("Failed to save product.");
  }

  return response.ok;
}

export async function updateProduct(productId: string, updatedProduct: ProductUpdate) {
  const response = await apiFetch(`/products/update/${productId}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(updatedProduct),
  });

  if (!response) {
    throw new Error("Failed to update product.");
  }

  return response.ok;
}

export async function deleteProductById(productId: string) {
  const response = await apiFetch(`/products/delete/${productId}`, {
    method: "DELETE",
  });

  if (!response) {
    throw new Error("Failed to delete product.");
  }

  return response.ok;
}
